### PR TITLE
SQL-817: Implement VAR and VARP as formulas

### PR DESCRIPTION
**Pull request by nbagnard**

### CODE REVIEW

File: connector/dialect.tdd [hunk 1]
```diff
@@ -145,6 +145,16 @@
       <unagg-formula>(CASE WHEN %1 IS NULL THEN NULL ELSE 0 END)</unagg-formula>
       <argument type='int' />
     </function>
+    <function group='aggregate' name='VAR' return-type='real'>
+      <formula>(STDDEV_SAMP(%1) * STDDEV_SAMP(%1))()</formula>
```
Comment (mattChiaravalloti) [inline connector/dialect.tdd]:
```suggestion
      <formula>(STDDEV_SAMP(%1) * STDDEV_SAMP(%1))</formula>
```
Are those `()` at the end intentional or a typo?

File: connector/dialect.tdd [hunk 2]
```diff
@@ -145,6 +145,16 @@
...
+      <unagg-formula>(CASE WHEN %1 IS NULL THEN NULL ELSE 0.0 END)</unagg-formula>
+      <argument type='real' />
+    </function>
+    <function group='aggregate' name='VARP' return-type='real'>
+      <formula>((STDDEV_POP(%1) * STDDEV_POP(%1)))</formula>
```
Comment (mattChiaravalloti) [inline connector/dialect.tdd]:
```suggestion
      <formula>(STDDEV_POP(%1) * STDDEV_POP(%1))</formula>
```
Is one set of parentheses surrounding the formula enough or do we need 2?

**Review by mattChiaravalloti — APPROVED** [review]
Looks good 👍 two small question-suggestions! I assume they can either be accepted or there is a valid reason for them, so either way -- no need to re-tag me!

File: connector/dialect.tdd (hunk 2 above)
Comment (nbagnard) [reply connector/dialect.tdd]: No, both are typos. 
I don't know how I did end up with these here 🤪
Fixed now, thank you for spotting them right away!

File: connector/dialect.tdd [hunk 3]
```diff
@@ -145,6 +145,16 @@
       <unagg-formula>(CASE WHEN %1 IS NULL THEN NULL ELSE 0 END)</unagg-formula>
       <argument type='int' />
     </function>
+    <function group='aggregate' name='VAR' return-type='real'>
+      <formula>(STDDEV_SAMP(%1) * STDDEV_SAMP(%1))</formula>
+      <unagg-formula>(CASE WHEN %1 IS NULL THEN NULL ELSE 0.0 END)</unagg-formula>
```
Comment (bucaojit) [inline connector/dialect.tdd]: I looked into this one and the unagg-formula should be NULL as the sample of a single value returns null.  There's also the existing STDEV function that can be switched to NULL on lines 130 and 135 if you'd like to make the change. 
https://www.mongodb.com/docs/manual/reference/operator/aggregation/stdDevSamp/#single-value

```suggestion
      <unagg-formula>NULL</unagg-formula>
```

**Review by bucaojit — CHANGES_REQUESTED** [review]
Great work finding that the functions can be implemented in the formulas.  

One comment about the unagg-formula.

File: connector/dialect.tdd (hunk 3 above)
Comment (nbagnard) [reply connector/dialect.tdd]: Yes, you are right.
I originally had NULL, but then I found this in [Tableau docs](https://tableau.github.io/connector-plugin-sdk/docs/dialect)   : 
> unagg-formula element The unagg-formula (unaggregated formula) element is an optional child of function elements. It should be specified only if the function is part of the aggregate group. Unaggregated formulas should represent a reasonable way of expressing the aggregate of a single value. For example, for average, it should be the value itself. **For variance, it should be 0.**
Looks like NULL makes more sense.
Changed it to NULL. Thanks for pointing this out!

**Review by bucaojit — APPROVED** [review]
LGTM

**Comment by nbagnard** [comment]
evergreen merge

**Pull request merged**